File: 백준/Silver/11722. 가장 긴 감소하는 부분 수열/가장 긴 감소하는 부분 수열.py
import sys

def solution():
  N = int(sys.stdin.readline())
  numbers = list(map(int, sys.stdin.readline().split()))
  dp = [1 for _ in range(N)]

  for i in range(1, N):
    for j in range(i):
      if numbers[i] < numbers[j]:
        dp[i] = max(dp[i], dp[j] + 1)
  
  print(max(dp))
    
    
solution()

File: 백준/Silver/11722. 가장 긴 감소하는 부분 수열/가장 긴 감소하는 부분 수열.js
let fs = require('fs');
let inputs = fs.readFileSync('/dev/stdin').toString().split('\n');

const N = Number(inputs[0]);
const numbers = inputs[1].split(' ').map(Number);
const dp = [1];

for (let i = 1; i < N; i++) {
    let maxLength = 0;
    for (let j = i - 1; j >= 0; j--) {
        if (numbers[j] > numbers[i] && maxLength < dp[j]) {
            maxLength = dp[j]
        }
    }
    maxLength === 0 ? dp.push(1) : dp.push(maxLength + 1);
}

console.log(Math.max(...dp));